### PR TITLE
[Az.Reservations] Add Az.Reservations cmdlets and tests

### PR DESCRIPTION
**Pull request by gaoyp830**
## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Reservations/README.md [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+> `autorest-beta`
+
+---
+### AutoRest Configuration
+> see https://aka.ms/autorest
+
+``` yaml
+require:
+  - $(this-folder)/../readme.azure.noprofile.md
+# lock the commit
```
Comment (BethanyZhou) [inline src/Reservations/README.md:52]: To lock the commit, could you please use `branch` tag with current commit id as value here?

File: src/Reservations/docs/Get-AzReservation.md [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+```
+
+Get the list of reservation orders in the current tenant.
+By design, some propeties do not have data due to the api response(e.g.
+LastUpdatedDateTime and SkuDescription).
+In this case please get the single reservation with command in example 2 to get the missing data.
+
+Some data might be trucated due to the width of powershell view, appending this to the end of the command to show the truncated data: | ft -Wrap
+
+### Example 2: {{ Add title here }}
```
Comment (BethanyZhou) [inline src/Reservations/docs/Get-AzReservation.md:63]: please fill this placeholder.

File: src/Reservations/docs/Get-AzReservationOrderId.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+Get-AzReservationOrderId -InputObject <IReservationsIdentity> [-DefaultProfile <PSObject>]
+ [<CommonParameters>]
+```
+
+## DESCRIPTION
+Get applicable `Reservation`s that are applied to this subscription or a resource group under this subscription.
+
+## EXAMPLES
+
+### Example 1: {{ Add title here }}
```
Comment (BethanyZhou) [inline src/Reservations/docs/Get-AzReservationOrderId.md:31]: Could you please check your examples and make sure all placeholders are filled?

**Review by BethanyZhou — CHANGES_REQUESTED** [review]
Generally looks good to me.

**Comment by BethanyZhou** [comment]
CI failed as handcrafted Az.Reservations exits in main branch. Need to remove it from main branch first.

File: src/Reservations/docs/Invoke-AzReservationsAvailableReservationScope.md [hunk 4]
```diff
@@ -0,0 +1,246 @@
+---
+external help file:
+Module Name: Az.Reservations
+online version: https://docs.microsoft.com/powershell/module/az.reservations/invoke-azreservationsavailablereservationscope
+schema: 2.0.0
+---
+
+# Invoke-AzReservationsAvailableReservationScope
```
Comment (BethanyZhou) [inline src/Reservations/docs/Invoke-AzReservationsAvailableReservationScope.md:8]: Is Invoke-AzReservation**_s_**AvailableReservationScope your expected name? Should it be Invoke-AzReservationAvailableReservationScope? Could you check?

File: src/Reservations/docs/Get-AzReservationOrderId.md (hunk 3 above)
Comment (gaoyp830) [reply src/Reservations/docs/Get-AzReservationOrderId.md]: Filled the missing titles

**Comment by gaoyp830** [comment]
> CI failed as handcrafted Az.Reservations exits in main branch. Need to remove it from main branch first.

Are we merging this PR to main branch after it's got checked in? I thought this is only for generation branch?

File: src/Reservations/docs/Invoke-AzReservationsAvailableReservationScope.md (hunk 4 above)
Comment (gaoyp830) [reply src/Reservations/docs/Invoke-AzReservationsAvailableReservationScope.md]: Good catch! I change the name of this command to `Get-AzReservationAvailableScopes`

File: src/Reservations/README.md (hunk 1 above)
Comment (BethanyZhou) [reply src/Reservations/README.md]: I think we need this tag to ensure each build can generate same code.

File: src/Reservations/docs/Invoke-AzReservationsAvailableReservationScope.md (hunk 4 above)
Comment (BethanyZhou) [reply src/Reservations/docs/Invoke-AzReservationsAvailableReservationScope.md]: Could you name it as Get-AzReservationAvailableScope as we need to keep singularity for naming cmdlet.

**Comment by BethanyZhou** [comment]
> Are we merging this PR to main branch after it's got checked in? I thought this is only for generation branch?

Our pipelines combine both branch together to analyze. Let's merge this PR to intermediate branch.

File: src/Reservations/docs/Invoke-AzReservationsAvailableReservationScope.md (hunk 4 above)
Comment (gaoyp830) [reply src/Reservations/docs/Invoke-AzReservationsAvailableReservationScope.md]: Changed the name to Get-AzReservationAvailableScope

File: src/Reservations/README.md (hunk 1 above)
Comment (gaoyp830) [reply src/Reservations/README.md]: Set branch to `main`

File: src/Reservations/README.md [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+## Run Generation
+In this directory, run AutoRest:
+> `autorest-beta`
+
+---
+### AutoRest Configuration
+> see https://aka.ms/autorest
+
+``` yaml
+branch: main
```
Comment (BethanyZhou) [inline src/Reservations/README.md:50]: Sorry for confusing you. The branch tag here is not used to describe actual branch. It means commit it actually. Its value should be from red box.
image

Comment (BethanyZhou) [inline src/Reservations/README.md:50]: Sorry for confusing you. The branch tag here is not used to describe actual branch. It means commit id actually. Its value should be from red box. The content of json file always changes even in main branch. We use this tag to ensure we can obtain same generated code each time.
image
```suggestion
branch: 01f2d2fc146e00485ecd150fc42bb2ee2abc400b
```

Comment (gaoyp830) [reply src/Reservations/README.md]: Thanks for the explanation

**Pull request merged**